sdk/core/azure-core/src/http/bearer_token_authentication_policy.cpp

```diff
@@ -102,7 +116,16 @@ void BearerTokenAuthenticationPolicy::AuthenticateAndAuthorizeRequest(
   // Check if token needs refresh for the second time in case another thread has just updated it.
   if (TokenNeedsRefresh(m_accessToken, m_accessTokenContext, currentTime, tokenRequestContext))
```

Comment:
Add a non-defaulted 5th parameter to `TokenNeedsRefresh()`: `bool forceRefresh`, then this line becomes
```suggestion
  if (TokenNeedsRefresh(m_accessToken, m_accessTokenContext, currentTime, tokenRequestContext, m_invalidateToken))
```
and the same happens to line 108.

The `TokenNeedsRefresh()` becomes the following:
```cpp
bool TokenNeedsRefresh(
    Azure::Core::Credentials::AccessToken const& cachedToken,
    Azure::Core::Credentials::TokenRequestContext const& cachedTokenRequestContext,
    Azure::DateTime const& currentTime,
    Azure::Core::Credentials::TokenRequestContext const& newTokenRequestContext,
    bool forceRefresh)
{
  return forceRefresh
      || newTokenRequestContext.TenantId != cachedTokenRequestContext.TenantId
      || newTokenRequestContext.Scopes != cachedTokenRequestContext.Scopes
      || currentTime > (cachedToken.ExpiresOn - newTokenRequestContext.MinimumExpiration);
}
```
Since bool value check is the simplest, we do that check first.

You also could not add the 5th parameter, but write the conditions as `if (m_invalidateToken || TokenNeedsRefresh(...)` on line 117 and `if (!m_invalidateToken && !TokenNeedsRefresh(...))` on line 108, but I think it is more error prone over time, as code evolves, to forget `m_invalidateToken` somewhere, or to mix up the cases when it needs to be negated or not. After all, if in the future there is a place where we genuinely want to ignore the value of `m_invalidateToken`, we can invoke it with 5th parameter, `forceRefresh = false`.